generic/tests/ping.py

```diff
@@ -5,129 +5,170 @@
+            status, output = _ping_with_params(ip, interface)
+            _check_ping_result(status, output)
 
-        if status != 0:
-            raise error.TestFail("Ping returns non-zero value %s" % output)
-
-        if ping_ext_host:
-            error.base_context("Ping test from guest side,"
-                               " dest: '%s'" % ext_host, logging.info)
-            pkt_sizes = packet_sizes
-            # There is no ping program for guest, so let's hardcode...
-            cmd = ['ping']
-            cmd.append(ext_host)  # external host
+    if ping_ext_host:
+        error_context.context("Ping test from guest side,"
+                              " dest: '%s'" % ext_host, logging.info)
 
+        for size in packet_sizes:
             if params.get("os_type") == "windows":
-                cmd.append("-n 10")
-                cmd.append("-l %s")
-                # Windows doesn't support ping with packet
-                # larger than '65500'
-                pkt_sizes = [p for p in packet_sizes if p < 65500]
-                # Add a packet size just equal '65500' for windows
-                pkt_sizes.append(65500)
+                logging.info("Ping with packet size: %s" % size)
+                status, output = _ping_with_params(packet_size=size,
+                                                   session=session)
+                _check_ping_result(status, output)
             else:
-                cmd.append("-c 10")  # ping 10 times
-                cmd.append("-s %s")  # packet size
-            cmd = " ".join(cmd)
-            for size in pkt_sizes:
-                error.context("Ping with packet size %s" % size,
+                for interval in interval_times:
+                    logging.info("Ping with packet size: %s and interval: %s" %
+                                 (size, interval))
+                    status, output = _ping_with_params(packet_size=size,
+                                                       session=session,
+                                                       interval=interval)
+                    _check_ping_result(status, output)
+
+        # ping with exceeded packages
+        error_context.context("Ping test with exceeded package",
                               logging.info)
-                status, output = session.cmd_status_output(cmd % size,
-                                                           timeout=60)
-                _get_loss_ratio(output)
-
-                if status != 0:
-                    raise error.TestFail(("Ping external host failed,"
-                                          " status: %s, output: %s" %
-                                          (status, output)))
+        status, output = _ping_with_params(packet_size=65508, session=session)
+        if output.find(result_str) == -1:
+            test.fail(("Ping test with exceeded package failed,"
+                       " expected string: '%s' not found" %
+                       result_str))
```

Comment:
This section appeared twice, although combined with different parameters. I suggest to move it out from the individual flows and keep only one.